sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c

```diff
@@ -0,0 +1,390 @@
+    dro->dodag_id = p2p_ext->dodag->dodag_id;
+    pkt = opt_snip;
+
+    if ((opt_snip = gnrc_icmpv6_build(pkt, ICMPV6_RPL_CTRL, GNRC_RPL_P2P_ICMPV6_CODE_DRO,
+                                      sizeof(icmpv6_hdr_t))) == NULL) {
+        DEBUG("RPL-P2P: cannot allocate ICMPv6 - no space left in packet buffer\n");
+        gnrc_pktbuf_release(pkt);
+        return NULL;
+    }
+
+    pkt = opt_snip;
+
+    return pkt;
+}
+
+void gnrc_rpl_p2p_send_DRO(gnrc_pktsnip_t *pkt, gnrc_rpl_p2p_ext_t *p2p_ext)
+{
+    assert(p2p_ext != NULL);
+
+    if (pkt == NULL) {
+        if ((pkt = _build_initial_DRO(p2p_ext)) == NULL) {
+            DEBUG("RPL-P2P: BUILD INITIAL DRO - no space left in packet buffer\n");
+            return;
+        }
+    }
+
+    gnrc_rpl_send(pkt, NULL, NULL, &p2p_ext->dodag->dodag_id);
+
+    return;
+}
+
+void gnrc_rpl_p2p_recv_DRO(gnrc_pktsnip_t *pkt, ipv6_addr_t *src)
+{
+    gnrc_pktsnip_t *icmpv6_snip = gnrc_pktbuf_mark(pkt, sizeof(icmpv6_hdr_t), GNRC_NETTYPE_ICMPV6);
+    gnrc_pktsnip_t *dro_snip = gnrc_pktbuf_mark(pkt, sizeof(gnrc_rpl_p2p_dro_t),
+                                                GNRC_NETTYPE_UNDEF);
+    gnrc_pktsnip_t *rdo_snip = gnrc_pktbuf_mark(pkt, sizeof(gnrc_rpl_p2p_opt_rdo_t),
+                                                GNRC_NETTYPE_UNDEF);
+    gnrc_pktsnip_t *addr_snip = pkt;
+    gnrc_rpl_instance_t *inst;
+    gnrc_rpl_dodag_t *dodag;
+    gnrc_rpl_p2p_ext_t *p2p_ext;
+    gnrc_rpl_p2p_dro_t *dro;
+    gnrc_rpl_p2p_opt_rdo_t *rdo;
+    uint8_t *addr_vec;
+    uint16_t flags;
+    size_t addr_len;
+
+    if (!rdo_snip || !dro_snip) {
+        DEBUG("RPL-P2P: Error - No DRO or RDO received\n");
+        gnrc_pktbuf_release(pkt);
+        return;
+    }
+
+    dro = dro_snip->data;
+
+    if ((inst = gnrc_rpl_instance_get(dro->instance_id)) == NULL) {
+        DEBUG("RPL-P2P: Error - Instance (%d) does not exist\n", dro->instance_id);
+        return;
+    }
+
+    dodag = &inst->dodag;
+
+    if ((p2p_ext = gnrc_rpl_p2p_ext_get(dodag)) == NULL) {
+        DEBUG("RPL-P2P: Error - No P2P-RPL DODAG extension found\n");
+        return;
+    }
+
+    if (p2p_ext->for_me) {
+        DEBUG("RPL-P2P: Ignore DRO\n");
+        return;
+    }
+
+    flags = byteorder_ntohs(dro->flags_rev);
+    p2p_ext->stop = (flags & (1 << GNRC_RPL_P2P_DRO_FLAGS_STOP)) >> GNRC_RPL_P2P_DRO_FLAGS_STOP;
+    p2p_ext->dro_ack = (flags & (1 << GNRC_RPL_P2P_DRO_FLAGS_ACK)) >> GNRC_RPL_P2P_DRO_FLAGS_ACK;
+    p2p_ext->dro_seq = (flags & (0x3 << GNRC_RPL_P2P_DRO_FLAGS_SEQ)) >> GNRC_RPL_P2P_DRO_FLAGS_SEQ;
+
+    addr_len = sizeof(ipv6_addr_t) - p2p_ext->compr;
+    ipv6_addr_t addr = p2p_ext->dodag->dodag_id;
+    ipv6_addr_t *me = NULL;
+    kernel_pid_t if_id;
+    addr_vec = addr_snip->data;
+
+    rdo = rdo_snip->data;
+
+    if (rdo->lmn > 0) {
+        rdo->lmn--;
+        memcpy(&addr.u8[p2p_ext->compr], &addr_vec[addr_len * rdo->lmn], addr_len);
+    }
+
```

Comment:
That was exactly the point where my mind-twists started :)
I read section 9.1 in rfc6997 and thought we shouldn't add a FIB entry for a hop-by-hop state to avoid interference.